### PR TITLE
Delete KnativeKafka resources when CR is deleted

### PR DESCRIPTION
**Pull request by aliok**
Bases https://github.com/openshift-knative/serverless-operator/pull/487

Similar to other PRs in the umbrella ticket, https://github.com/openshift-knative/serverless-operator/issues/486, I didn't want to wait until the CI is green and that's merged. This is a WIP PR to see in advance if things fail and if possibly get some reviews.

This PR deletes the resources when the KnativeKafka CR is deleted.
Specifically,
- KafkaChannel resources --> garbage collection also works for these since they're in the same namespace and we set ownerRefs for them
- KafkaSource resources that are put in another namespace, `knative-sources` -->  no GC, manual deletion for these
- Deletion of cluster-scoped resources

Deletion of the resources when a component gets disabled is to be handled in another PR (e.g. `spec.channel.enabled=true` is changed to `spec.channel.enabled=false`)

As written in the umbrella ticket,  https://github.com/openshift-knative/serverless-operator/issues/486, there's still a lot to do.

Verification:

1. Enable the controller by uncommenting the stuff in `knative-operator/pkg/controller/add_knativekafka.go`

2. Install operator and eventing by `make install`

3. Need to manually create the CRD since it is not in the CSV yet
```
kubectl apply -f knative-operator/deploy/crds/operator_v1alpha1_knativekafka_crd.yaml
```

4. Install:
```
cat <<EOS |kubectl apply -f -
---
apiVersion: operator.serverless.openshift.io/v1alpha1
kind: KnativeKafka
metadata:
  name: example
  namespace: knative-eventing
spec:
  source:
    enabled: true
  channel:
    enabled: true
    bootstrapServers: foo.example.com:1234
...
EOS
```

5. Delete the CR and check if created resources are deleted:
```
kubectl delete knativekafkas example -n knative-eventing
```

6. Following commands should say `Error from server (NotFound): error when deleting "knative-operator/deploy/resources/knativekafka/kafka....yaml": foo "bar" not found` for each of the resource defined in them
```
k delete -f knative-operator/deploy/resources/knativekafka/kafkasource-latest.yaml
k delete -f knative-operator/deploy/resources/knativekafka/kafkachannel-latest.yaml
```

### CODE REVIEW

File: knative-operator/deploy/resources/knativekafka/README.md
Comment (aliok) [inline knative-operator/deploy/resources/knativekafka/README.md:1]: Simple instructions to download the KafkaChannel and KafkaSource manifests. Not really relevant to this PR, but let's merge this too.

File: knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.4.yaml
Comment (aliok) [inline knative-operator/deploy/resources/knativekafka/kafkachannel-v0.17.4.yaml:1]: Removed all the comments and blank lines in the manifests

File: knative-operator/deploy/resources/knativekafka/kafkasource-v0.17.4.yaml
Comment (aliok) [inline knative-operator/deploy/resources/knativekafka/kafkasource-v0.17.4.yaml:1]: Removed all the comments and blank lines in the manifests

File: knative-operator/pkg/apis/operator/v1alpha1/knativekafka_lifecycle_test.go
Comment (aliok) [inline knative-operator/pkg/apis/operator/v1alpha1/knativekafka_lifecycle_test.go:1]: As requested in the previous PR, removed the upstream license headers

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:135]: Now we're actually setting the finalizer on the CR.
This is necessary to be able to remove resources that can't be GC'ed because they're in a different namespace (`knative-sources` namespace)

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: setting the ownerRef for the resources here for the KafkaChannel resources (which are in the same namespace with the CR)

File: knative-operator/deploy/resources/knativekafka/README.md
Comment (matzew) [reply knative-operator/deploy/resources/knativekafka/README.md:1]: Not sure I get the value here, tbh

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (matzew) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:29]: what earlier version? you mean the `knative-kafka` operator?

File: knative-operator/deploy/resources/knativekafka/README.md
Comment (aliok) [reply knative-operator/deploy/resources/knativekafka/README.md:1]: Well, there's a `sed` command to delete comments and blank lines from manifests.
That's all :)

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:29]: no, that's copy paste :)

but not a bad one actually. it means, don't change it to something else in the future!

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (matzew) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:135]: IMO it would be nicer, if we just stash them all to the `knative-eventing` namespace.

I will propose the same upstream - insted of the `knative-sources` namespace ,....

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (matzew) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:29]: OK, I'd perhaps do that explicit

**Comment by matzew** [comment]
so. this looks good, and similar to related code here

/lgtm

I think @markusthoemmes is approver for final blessing :shrug:

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:135]: well, we still need the finalizer, so, this code won't change.
There are also cluster scoped resources which can't be GC'ed and we still need the finalizer to prevent the deletion of KnativeKafka CR before them.

Happy to discuss the get rid of `knative-sources` to `knative-eventing` for KafkaSource.

IMO we should do that change separately once everything is working. It might require some upstream changes perhaps, I don't know yet.

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:30]: Updated now

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: Do we need the intermediary slice? Can't we just put the `mf.InjectOwner(instance)` call into the Transform call below?

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (markusthoemmes) [inline knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:275]: copy and paste issue.

File: knative-operator/deploy/resources/knativekafka/README.md
Comment (markusthoemmes) [reply knative-operator/deploy/resources/knativekafka/README.md:1]: Seems odd to dump on this PR but :man_shrugging:

**Review by markusthoemmes — COMMENTED** [review]
Small nits, then LGTM

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: nope, we don't need it. let me get rid of that.

Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]: Done now

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go]: fixed, thx

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: That's why the owner refs are only set for the channel resources, which live in `knative-eventing` namespace (same namespace as the CR).
It is not set for the KafkaSource resources, which live in `knative-sources` namespace.

> Otherwise if I create this in namespace foo, how are the owner refs going to work?

You can't create the KnativeKafka CR in a namespace other than `knative-eventing`. This is similar to existing CRs for serving and eventing. That check is done in another PR (https://github.com/openshift-knative/serverless-operator/pull/510/commits/d2895c8729793c98c9bfcd78606ea1f3d6b1865a#diff-186644545bff629e539219c7f924fc96R111) already but there's a lot of stuff to merge until that check is enabled.

Comment (markusthoemmes) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: Hmm, ok.

Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: @matzew has some thoughts about using `knative-eventing` for KafkaSource as well, but I don't want to do this until I merge my existing work.

Also, not sure if it requires any changes in the upstream code.

Comment (matzew) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: @aliok let's track it

**Comment by matzew** [comment]
@markusthoemmes wanna do the final blessing ?

**Review by markusthoemmes — APPROVED** [review]
/lgtm
/approve

File: knative-operator/pkg/controller/knativekafka/knativekafka_controller.go
Comment (aliok) [reply knative-operator/pkg/controller/knativekafka/knativekafka_controller.go:200]: @matzew already added in the umbrella ticket: https://github.com/openshift-knative/serverless-operator/issues/486

**Pull request merged**
